### ISSUE TITLE
Next13 caching fetched data even with the 'revalidate = 0' route segment

### ISSUE THREAD

**Issue by EvandrooViegas** [vercel/next.js#51788]
### Verify canary release

- [x] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: win32
      Arch: x64
      Version: Windows 10 Home
    Binaries:
      Node: 18.12.1
      npm: N/A
      Yarn: N/A
      pnpm: N/A
    Relevant Packages:
      next: 13.4.8-canary.2
      eslint-config-next: 13.4.6
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.1.3
    Next.js Config:
      output: N/A
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App directory (appDir: true), Data fetching (gS(S)P, getInitialProps)

### Link to the code that reproduces this issue or a replay of the bug

https://github.com/EvandrooViegas/outsystem-angola

### To Reproduce

1. clone my [repo](https://github.com/EvandrooViegas/outsystem-angola)
2. run `npm install` in the root of the cloned repo
3. run `npm run dev`

### Describe the Bug

Next13 seems to be caching the data coming from supabase even with the `revalidate = 0` route segment.

### Expected Behavior

Get the freshest data in every page request

### Which browser are you using? (if relevant)

chrome

**Comment by AlexHappyCode** [comment]
im having the same issue. I tried all of these

```
export const fetchCache = 'force-no-store'
export const revalidate = 0 // seconds
export const dynamic = 'force-dynamic'
```

I looked in the network tab and i see this response header:

`X-Nextjs-Cache: HIT
`

The docs talk about the X-Nextjs-Cache for images but its applied to the route as a whole in my case

**Comment by kevincobain2000** [comment]
Doesn't work, it keeps fetching the cached data

`"next": "13.4.7",`

```js
export const fetchCache = 'force-no-store'
export const revalidate = 0 // seconds
export const dynamic = 'force-dynamic'

export default async function GetData() {
    const data = await fetch("http://127.0.0.1:3001/time", {
        next: { revalidate: 0 },

    })

    const json = await data.json()

    console.log(json)
    return (
        <div>
            <ul>
                {json.time}
            </ul>
        </div>
    )
}
```

**Comment by kevincobain2000** [comment]
https://www.reddit.com/r/nextjs/comments/12ih2gc/revalidate_not_working_on_deployed_website_using/jq5gn2f/?utm_source=share&utm_medium=ios_app&utm_name=ioscss&utm_content=1&utm_term=1&context=3

Possible fix.

**Comment by nevenhsu** [comment]
Same here.
I currently add a random number as query to bypass this bug.
Something like: http://localhost:3000/?**t=1688177362**

**Comment by kevincobain2000** [comment]
Upon little more investigation, `revalidate:0` works but only for one particular use.
If the component that is fetching the data is server `use server` then upon hard refresh, it fetches each time, but when navigating via `<Link` it uses the old cache and never hits a request.

**Comment by naishe** [comment]
Linking it to an open discussion https://github.com/vercel/next.js/discussions/51612#discussion-5323305 

@nevenhsu -- unfortunately, this does not solve the page redirection issue or browser back button issue. So, I have a detail page `/address/[id]` and an edit page `/address/[id]/edit`. When the form on the edit page gets submitted, I redirect them back to `/address/${id}?r=${Math.random()}`. You see the new data on this URL with new query string, *BUT* it does not clear the page cache. If you go to `/address/[id]` from menu which uses `<Link href=`/address/${id}`>...`, you still see this old data. So, either you always uglify your URL with `Math.random()` all the time and never use browser's back button, or convert it to client component.

I ended up using `page` as just static shell and moved on to React-Query. The peace of mind of knowing that you are in control and not some voodoo magic keyword is returning data based on its mood, is unparalleled.

**Comment by kevincobain2000** [comment]
Thanks @naishe 
I mean it’d be nice if nextjs fixes it, as this is a much desirable feature to unleash power of react server side components. 
A bit disappointed that they pushed half baked on the major version

**Comment by EvandrooViegas** [comment]
> Doesn't work, it keeps fetching the cached data
> 
> `"next": "13.4.7",` Edit* Also in `13.4.8-canary.14`
> 
> ```js
> export const fetchCache = 'force-no-store'
> export const revalidate = 0 // seconds
> export const dynamic = 'force-dynamic'
> 
> export default async function GetData() {
>     const data = await fetch("http://127.0.0.1:3001/time", {
>         next: { revalidate: 0 },
> 
>     })
> 
>     const json = await data.json()
> 
>     console.log(json)
>     return (
>         <div>
>             <ul>
>                 {json.time}
>             </ul>
>         </div>
>     )
> }
> ```

If that is a component make sure to add the `export const revalidate = 0` and `export const dynamic = 'force-dynamic'` route segments in the page that is rendering that component

**Comment by kevincobain2000** [comment]
Yes, thanks @EvandrooViegas 
I tried with route segments, layout and even putting the fetch code inside the route component as well.
It only works when I manually refresh the page. Upon route Link click, it never hits the GET request.

**Comment by EvandrooViegas** [comment]
> Yes, thanks @EvandrooViegas I tried with route segments, layout and even putting the fetch code inside the route component as well. It only works when I manually refresh the page. Upon route Link click, it never hits the GET request.

could you provide the repo link?

**Comment by Vinlock** [comment]
I'm getting this same issue. Nothing is working besides hard refreshing or `revalidatePath()` which feels so far removed...

**Comment by matannahmani** [comment]
this is literally the worst thing, why release it if it's not working...
back button / link always keep cache no matter what you do

**Comment by EvandrooViegas** [comment]
I found this [video](https://www.youtube.com/watch?v=v_HpVwCDAgU&ab_channel=DaveGray) very helpful

**Comment by jack-szeto** [comment]
According to the video that @EvandrooViegas shared. I got this work. I'm sharing this for my future self and hope this can help someone. And point me out if I'm wrong.

```tsx
// components/post-editor.tsx
"use client"
import React, { startTransition, useCallback } from "react";
import { useRouter } from "next/navigation";
// ...

export const PostEditor = ({ post }: { post: Post }) => {
    const router = useRouter();
    const updatePost = useCallback(async (values: Post) => {
        // I'm using ky to fetch data, I guess you can use native fetch() as well
        await ky
            .put(`/api/post/${values.id}`, {
                json: values,
            });
        startTransition(() => {
            router.refresh();
        });
    }, [router]);

    return (
        // ...
    );
}
```
```tsx
// /app/post/[postId]/page.tsx
const Page = async ({
    params,
}: {
    params: { postId: string };
    searchParams: { [key: string]: string | string[] | undefined };
}) => {
    const post = await GetPost(params.postId);

    return (
        <PostEditor post={post} />
    )
}
export default Page;
```

Basically, I'm using `startTransition` to call the `router.refresh()` whenever I update the posts. This can refresh the pages cached by Router without affecting the UI.

**Comment by kevincobain2000** [comment]
Thanks for the video @EvandrooViegas  and solution @jack-szeto 
My take away from the video is that Nextjs team is working on it. Ideally we shouldn't have to add `startTransition`, `route.refresh` on each time we want to make a navigation work.

**Comment by masterkain** [comment]
I get this too -- during an investigation on why caching works the way it does on a kubernetes cluster I get some unexpected behavior.

navigation on simple sidebar `Link`s is being served from somewhere for an unspecified amount of time, I think this is browser behavior.

navigating to a `revalidate = 0` page generates this response, which is clearly trying its best to avoid caching

<img width="581" alt="Screenshot 2023-07-09 at 12 24 27" src="https://github.com/vercel/next.js/assets/12844/1c1e605e-c8c6-4bc6-af54-ffd2c8cbdbd9">

yet when revisiting the same link there are no requests generated. only after an unspecified amount of time the browser understands it needs to ask again the resource to the server.

compared that kind of response to one generated from a generic rails server the only thing that stands out is the presence of a weak etag

**Comment by matan00** [comment]
@masterkain apparently the client side router doesn't respect this behavior therefore links and back button doesn't work as expected to disable this functionality, you have to put on the <Link prefetch={false}> this will solve the problem most likely, but then again this solution in my opinion is not good enough, especially when you're doing a big batch of calls on server components everytime you have to refetch it.
My solution was integrating tanstack query on client side and on server component I dehydrate the query to the client, it's not perfect, it doesn't feel good, but it's what it's for now, hopefully next team will make the client router behavior more dynamic and allow us more control instead of all of this magic.

**Comment by masterkain** [comment]
this has also some interesting details and updates https://github.com/vercel/next.js/issues/42991

**Comment by hozunlee** [comment]
<img width="577" alt="image" src="https://github.com/vercel/next.js/assets/60101732/6b9b8cf2-31f8-4d17-a30e-61713d1796ba">

x-vercel-cache: HIT  xxxxxx

nextjs + vercel

The fetch works properly locally, but there is a problem when I distribute it to the vercel.

'/' The route only gets new data when redistributed, and when you reload the page, 304 appears.

Funny thing is that '/post/[id]' becomes a prerender and gets the data right.

x-vercel-cache: HIT

I always want to get an x-vercel-cache: Miss.

Has anyone solved it?

**Comment by jandk008** [comment]
Got similar issue with nextjs + vercel and the fetched data is always cached no matter what solutions avoiding cache that is written in the document is used. 
But it works well locally.
And the nextjs + vercel works well after I downgraded nextjs to `13.3.4` from `13.4.7`. 
even with the sharing of reworked cache since `13.4.0` as someone else mentioned,  it looks odd to have unfunctioning `13.4.0` versions.

**Comment by knightrun** [comment]
im having the same issue The page should display new data every time it's accessed, but page.tsx keeps showing cached data without making API calls. I have tested it in the development environment. Is there any solution available to resolve this problem?

`"next": "13.4.9"`

I tried the following configurations, but unfortunately, they didn't work
`export const dynamic = 'force-dynamic';
export const fetchCache = 'force-no-store';
export const revalidate = 0;
cache: 'no-store'
`

**Comment by tegarjgap** [comment]
> im having the same issue The page should display new data every time it's accessed, but page.tsx keeps showing cached data without making API calls. I have tested it in the development environment. Is there any solution available to resolve this problem?
> 
> `"next": "13.4.9"`
> 
> I tried the following configurations, but unfortunately, they didn't work `export const dynamic = 'force-dynamic'; export const fetchCache = 'force-no-store'; export const revalidate = 0; cache: 'no-store' `

@knightrun I was add the same configuration, and work in development..
`export const fetchCache = "force-no-store";
export const revalidate = 0; // seconds
export const dynamic = "force-dynamic";`

you can look at it on my [repo](https://github.com/tegarjgap/nextjs-starter/blob/main/src/app/products/page.tsx)

**Comment by kevincobain2000** [comment]
Update: 
Doesn't work with `v13.4.11` https://github.com/vercel/next.js/releases/tag/v13.4.11

**Comment by knightrun** [comment]
@tegarjgap Thank you for your comment.

Have you verified if the `getAllProduct` is triggered when accessing the page using next/link or router push? It's still not working for me.

**Comment by ml-sketch** [comment]
I am having the same problem. I've tried it all, but it just won't work. I believe Vercel needs to fix this. It only works with router.refresh but that should be a big nono

**Comment by tegarjgap** [comment]
> @tegarjgap Thank you for your comment.
> 
> Have you verified if the `getAllProduct` is triggered when accessing the page using next/link or router push? It's still not working for me.

yeah @knightrun .. in my local development.. the CACHE - MISS show when call `getAllProduct()`

**Comment by tolgaand** [comment]
I've create a basic API, but I'm having a problem. Every time I send a request to the API, I keep getting the same URL back.

```javascript
import { NextResponse } from "next/server";

const images = [
  "https://i.imgur.com/x.png",
  "https://i.imgur.com/x1.png",
  "https://i.imgur.com/x2.png",
  "https://i.imgur.com/x3.png",
  "https://i.imgur.com/x4.png",
  "https://i.imgur.com/x5.png",
  "https://i.imgur.com/x6.png",
  "https://i.imgur.com/x7.png",
  "https://i.imgur.com/x8.png",
];

export async function GET() {
  const randomImage = images[Math.floor(Math.random() * images.length)];

  return NextResponse.json({
    image: randomImage,
  });
}
```

page.tsx:
```javascript
export const revalidate = 0;
export const dynamic = "force-dynamic";
```

and button onclick function:
```javascript
 const { image } = (
  await fetch(`/api/generate-nft-url?d=${Date.now()}`, {
    next: { revalidate: 0 },
    cache: "no-cache",
  })
).json();

startTransition(() => {
  router.refresh();
});
```

I will try all the suggestions above, but the problem doesn't get fixed.

**Comment by kevincobain2000** [comment]
Haven't tested it but https://github.com/vercel/next.js/releases/tag/v13.4.13-canary.8 seems promising for related issue here https://github.com/vercel/next.js/pull/53373

**Comment by emrecoban** [comment]
I've the same issue. Interesting problem, `revalidate` solve the problem in local, but not in prod.
```javascript
export const revalidate = 0;

// no need these ones, but you can try.
export const fetchCache = 'force-no-store'
export const dynamic = 'force-dynamic'
```
Is there any solution?
- Edge runtime
- Server component
- Supabase
- `/app/[username]/page.js`

**Comment by kevincobain2000** [comment]
It seems like the client side caching cannot be disabled https://youtu.be/_yhSh4g0NSk

**Comment by AnirudhSinghBhadauria** [comment]
Router cache cannot be neglected, sad.

**Comment by brain1401** [comment]
gyus, is this still unsolved  ?

**Comment by diveddie** [comment]
> 

I can confirm that @jack-szeto's recommendation to use `router.refresh()` works for me. I feel like this is a workaround for now, and I'd like to see that the server-side cache can be invalidated. Might just switch to React Query for the time being until this works 😐

**Comment by nodegin** [comment]
same here... as an user, we have no control on how next controls our cache?

**Comment by nodegin** [comment]
I found a solution that works for me (Apollo with RSC).

Just add the `middleware.ts` on your root directory with the following content:

```ts
import {
  NextRequest,
  NextResponse,
} from 'next/server'

export function middleware(request: NextRequest) {
  // Fix Next overaggressive caching
  return NextResponse.rewrite(request.url + `?${ Math.random() }`)
}
```

This ensures Next always fetch up-to-date data without cache internally.

**Comment by zhouw** [comment]
There’s a new client side cache implemented since v13.3.2 #48695

 When you use next/link to do soft navigation, dynamic pages are cached for 30seconds and there is no way to opt out. There has been discussion in other threads going on.

From the doc:

> [Opting out](https://nextjs.org/docs/app/building-your-application/caching#opting-out-3)
>
>It's not possible to opt out of the Router Cache.
>
> You can opt out of prefetching by setting the prefetch prop of the <Link> component to false. However, this will still temporarily store the route segments for 30s to allow instant navigation between nested segments, such as tab bars, or back and forward navigation. Visited routes will still be cached.

**Comment by akshit-mdr** [comment]
I am fetching data on server for `/user` route, now, I have `/user/edit` to edit the data of user which is also a server side component. When the data in edit route is updated, user is redirected back to `/user` route but even when using no-store, getting cached data only.

not able to use router.refresh() since all the entities involved are on server side.

What can be done here - 
1. Should I move all components to client side?
2. is there any tagging and revalidation method for achieving this?

Further, I am integrating zustand store with next 13 server side components, and not able to fetch state in server, so, have to make api call without checking if the data is already present in store. How can we take advantage of global store with next 13 server component?

Can someone help here, thanks

**Comment by lmf-git** [comment]
I'm going to wildly speculate that this is due to Next/Vercel not wanting to send so many requests from their side.

**Comment by ghcpuman902** [comment]
I've been having issues with caching as well but managed to solve now:

if the fetched source / API also has cache mechanism like stale-while-revalidate, when the page read subsequent fetch results as not changing, it will cache and not try to call the API again. 

Keep your API end point dynamic and not cached, and cache at the root level of your app to solve this issue. 

Basically don't revalidate = 3600 on both your page.js AND then fetch in page.js AND your route.ts, best to make page.js dynamic with searchParams, use next: { revalidate: 3600 } in your fetch, then make sure your API route is complete dynamic (revalidate = 0 works for me but if not try using request object.

If people want to see some code please let me know, I'll try to add them when I'm done travelling on this train.

**Comment by nevenhsu** [comment]
> I've been having issues with caching as well but managed to solve now:
> 
> if the fetched source / API also has cache mechanism like stale-while-revalidate, when the page read subsequent fetch results as not changing, it will cache and not try to call the API again.
> 
> Keep your API end point dynamic and not cached, and cache at the root level of your app to solve this issue.
> 
> Basically don't revalidate = 3600 on both your page.js AND then fetch in page.js AND your route.ts, best to make page.js dynamic with searchParams, use next: { revalidate: 3600 } in your fetch, then make sure your API route is complete dynamic (revalidate = 0 works for me but if not try using request object.
> 
> If people want to see some code please let me know, I'll try to add them when I'm done travelling on this train.

Please share some code with us, thx a lot!

**Comment by louispelarrey** [comment]
Still have this issue after upgrading to next.js 14.0.2

**Comment by kirkstrobeck** [comment]
My issue turned out to be a race condition that I was attributing to cache

**Comment by kevincobain2000** [comment]
On nesxtjs14, tags work!

**Comment by mirano-galijasevic** [comment]
Ok, so how can we turn off the damn caching in the end? I tested with 14.0.2 and it is not working at all, nothing has changed.
This practically renders the whole framework unusable. I cannot understand what the thinking behind this was, who would ever go and weld in the caching into the framework and saying: "hey, let's be faster, never mind if data will be wrong, who cares about that!".
The caching should be turned off by default, and then you configure it if you want it, not the other way around.
So, what are you doing to go around this? I see the hacks with adding bogus search params, revalidate, force-dynamic and other hacks do not work, so what do you do? Converting all components to client? Going down to 13.3.4 and then re-writing all the code that will not be working? What? Since nobody is even mentioning that this is being fixed.

**Comment by Rodo1foBandeira** [comment]
**In the Next 14 too.**
I tried several ways in the component
`app/pessoa/page.tsx`
```
import Link from 'next/link';
import { getPessoas } from '../../services/pessoa'

export const dynamic = 'force-dynamic';
export const revalidate = 0;
export const fetchCache = 'force-no-store';
export const runtime = 'edge';

export default async function Pessoas(){
    const tst= await getPessoas();
    return (
        <ul>
            { tst.map(x => (<li key={x.nome}><Link href={`/pessoa/${x.id}`}>{x.nome}</Link></li>))}
        </ul>
    )
}
```
In the fetch that my component consumes
`src/services/pessoa.tsx`
```
import { Pessoa } from "@/models/pessoa";

export async function getPessoas (): Promise<Pessoa[]> {
    const reponse = await fetch('http://localhost:3004/pessoa',
        {
            cache: "no-store"
        }
    );
    return await reponse.json();
}
```

**Comment by hozunlee** [comment]
> <img alt="image" width="577" src="https://private-user-images.githubusercontent.com/60101732/252150975-6b9b8cf2-31f8-4d17-a30e-61713d1796ba.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE3NjI5OTgsIm5iZiI6MTcwMTc2MjY5OCwicGF0aCI6Ii82MDEwMTczMi8yNTIxNTA5NzUtNmI5YjhjZjItMzFmOC00ZDE3LWEzMGUtNjE3MTNkMTc5NmJhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA1VDA3NTEzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVjNjFmNjdlOGM0Yzg5ZWNlZjM0OWRjYmE5ODVlODAwY2ZjOTk1NDhlOTMwM2Y1MzA1NGExYjk2NDk2NWRmMGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.5HXTzwdi_6gVRbYCfv_9cwcFjyR3VH9vS1H5Kfi9dfg">
> x-vercel-cache: HIT xxxxxx
> 
> nextjs + vercel
> 
> The fetch works properly locally, but there is a problem when I distribute it to the vercel.
> 
> '/' The route only gets new data when redistributed, and when you reload the page, 304 appears.
> 
> Funny thing is that '/post/[id]' becomes a prerender and gets the data right.
> 
> x-vercel-cache: HIT
> 
> I always want to get an x-vercel-cache: Miss.
> 
> Has anyone solved it?

https://hololog.dev/post/21
이걸로 해결했습니다.

**Comment by Rodo1foBandeira** [comment]
**Don't test in a browser**, use postman or insomnia. Note that it works normally. **The problem really is the browser!**

**Comment by ChoaibMouhrach** [comment]
Same issue 14.O.4

**Comment by fmsf** [comment]
Was  hitting the same issue yesterday. Just documenting my findings in case it helps other people. I tried multiple different combos of:
```
export const fetchCache = 'force-no-store'
export const revalidate = 0 // seconds
export const dynamic = 'force-dynamic'
```

And in the network headers I could see that vercel was applying the header to the requests comming from the browser. However my cache hits where happening between the server function and my external server.

My guess is that this was happening:

`Browser ---- (no cache applies correctly) -->  serverless function that does api calls `
then 
`serverless function that does api calls --- (no cache is no longer applied because I only requested it on the front end) ---> external API`

 Browsing among the many threads with the common topic of "Disabling cache doesn't work" I eventually found this one: Fetch caching in Server actions: https://github.com/vercel/next.js/discussions/50045 which pointed to two actions:

1 - Upgrade to next 14
2 - Use the new `unstable_noStore` to disable caching in individual functions within functions within a `.ts|.tsx` file tagged with `"use server"`.  https://nextjs.org/docs/app/api-reference/functions/unstable_noStore

This has worked for me. I hope it helps others.

**Comment by ChoaibMouhrach** [comment]
A solution I found is using  `unstable_noStore` with `router.refresh`

```ts

import { Suspense } from "react";
import { Table } from "./table";
import { unstable_noStore } from "next/cache";

const Products = () => {
  unstable_noStore();

  return (
    <Suspense fallback="loading...">
      <Table />
    </Suspense>
  );
};

export default Products;

```

```ts

"use client";

import { Button } from "@/components/ui/button";
import { usePathname } from "next/navigation";
import { useMemo } from "react";
import { useRouter } from "next/navigation";

interface Navigation {
  name: string;
  href: string;
}

const navigations: Navigation[] = [
  {
    name: "Dashboard",
    href: "/",
  },
  {
    name: "Products",
    href: "/products",
  },
  {
    name: "Sales",
    href: "/sales",
  },
  {
    name: "Sales Returns",
    href: "/sales-returns",
  },
  {
    name: "Suppliers",
    href: "/suppliers",
  },
  {
    name: "Purchases",
    href: "/purchases",
  },
  {
    name: "Purchases Returns",
    href: "/purchases-returns",
  },
  {
    name: "Settings",
    href: "/settings",
  },
];

interface SideBarItemProps {
  navigation: Navigation;
  pathName: string;
}

const SideBarItem: React.FC<SideBarItemProps> = ({ navigation, pathName }) => {
  const router = useRouter();

  const active = useMemo(() => {
    let active = false;

    if (`${pathName}/`.startsWith(`${navigation.href}/`)) {
      active = true;

      if (navigation.href === "/" && pathName !== "/") {
        active = false;
      }
    }

    return active;
  }, [navigation.href, pathName]);

  const navigate = () => {
    router.push(navigation.href);
    router.refresh();
  };

  return (
    <Button
      className="justify-start"
      variant={active ? "secondary" : "ghost"}
      onClick={() => navigate()}
      size="sm"
    >
      {navigation.name}
    </Button>
  );
};

export const SideBar = () => {
  const pathName = usePathname();

  return (
    <div className="w-72 shrink-0 border-r p-4 flex flex-col gap-4">
      {navigations.map((navigation, index) => (
        <SideBarItem pathName={pathName} navigation={navigation} key={index} />
      ))}
    </div>
  );
};
```

**Comment by realrufans** [comment]
I had the same issue with my  code. In my case, I wasn't  making use of Link component. So, users aren't expected to navigate from one page to the other. but still, my api data was being cached in page.tsx.

Here's what i did to disable the cache. or at least prevent if from returning stale data from api call on every call.
 
 I added ```export const revalidate = 0;```  at the top level of my route.ts file in the api folder.

**Comment by jovanhartono** [comment]
crazy that it's still an issue...

**Comment by salman0ansari** [comment]
this is so annoying

**Comment by nisan1337** [comment]
This makes app router unusable anything other than static pages. There is always a hydration error in dynamic pages that force root to client rendering which is basically plain react.

**Comment by ChoaibMouhrach** [comment]
Using "router.refresh" worked for me but I don't like how my progress bar stoped working too

**Comment by samcx** [comment]
Hi everyone, will be taking a look at this!

It seems the original <picture data-single-emoji=":repro:" title=":repro:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/repro/5cea53b5532825b4.png" alt=":repro:" align="absmiddle"></picture> that provided is currently a `404` (not a <picture data-single-emoji=":public-big:" title=":public-big:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/public-big/55b6a09c19a08dc0.png" alt=":public-big:" align="absmiddle"></picture> <picture data-single-emoji=":repro:" title=":repro:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/repro/5cea53b5532825b4.png" alt=":repro:" align="absmiddle"></picture> or doesn't exist anymore).

If anyone is able to provide a minimal <picture data-single-emoji=":repro:" title=":repro:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/repro/5cea53b5532825b4.png" alt=":repro:" align="absmiddle"></picture> , that will be great so we can confirm and fix this long-standing issue.

**Comment by ChoaibMouhrach** [comment]
@samcx How do we opt-out of client-side caching?

**Comment by samcx** [comment]
@ChoaibMouhrach I believe opting-out of client-side caching is related to the original issue above because the client-side cache is [different from the cache used by fetch](https://nextjs.org/docs/app/building-your-application/caching#apis). The only way to currently do this is using `router.refresh`.

Can you clarify why you wish to opt-out of client-side cache?

**Comment by ChoaibMouhrach** [comment]
@samcx In my case, I have a sidebar with links. When I navigate between those links, I retrieve fresh data from the database. However, when I create a new item in the database and then navigate back to the page via a link component, I don't see the newly created item. This is an issue for me. The only solution I found is using a router.refresh, but I encountered some drawbacks, one of which is the loss of my page navigation progress bar.

**Comment by samcx** [comment]
@ChoaibMouhrach My apologies, I take back what I said earlier—`revalidatePath` is indeed another option purge Client-side Router Cache.

In your use-case, you should use `revalidatePath` (https://nextjs.org/docs/app/api-reference/functions/revalidatePath) instead of `router.refresh`. Are you using a [Server Action](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations) to retrieve fresh data from the database?

I believe this will not cause the loss of your page navigation progress bar.

**Comment by ChoaibMouhrach** [comment]
@samcx Yes I'm using server actions to get the data, I think I tried it before and it didn't work, I'll give it a shot then I'll let you know

**Comment by ChoaibMouhrach** [comment]
@samcx I just had the time to try it, unfortunately, it didn't work, I still get the old data.

**Comment by samcx** [comment]
@ChoaibMouhrach Thanks for testing!

That seems like possibly a separate issue with `revalidatePath`. Can you submit a new [bug](https://github.com/vercel/next.js/issues/new?assignees=&labels=template:+bug&projects=&template=1.bug_report.yml) so we can take a look?

**Comment by ChoaibMouhrach** [comment]
@samcx, Thanks, I will

**Comment by jack-szeto** [comment]
hey it's been a while! I've been doing some testing and I think the issue **isn't a problem anymore**. Turns out everything works just fine when `unstable_cache` and `revalidateTag` are used correctly.

I've been experimenting and testing this in my own repo, [nextjs-caching-with-prisma](https://github.com/jack-szeto/nextjs-caching-with-prisma). If you've got some time, feel free to clone it and give it a shot.

To give you a quick rundown, here's what I've done. 
In the `src/lib/post.ts` file:
```ts
// ...
import { unstable_cache as cache, revalidateTag } from "next/cache";

export const getPosts = cache(
  async () => {
    return await prisma.post.findMany({
      // ...
    });
  },
  ["posts"],
  {
    tags: ["posts"],
  }
);

export const createPost = async (data: Prisma.PostCreateInput) => {
  const newPost = await prisma.post.create({
    data,
  });
  revalidateTag("posts");
  return newPost;
};

// ...
```
Then, I separated the server action and form validation logic in the `src/actions/post.action.ts` file:
```ts
"use server"; // server actions have to have the "use server" directive

export const createPostAction = async (prevState: any, formData: FormData) => {
  // validation

  const newPost = await createPost({
    // ...
  });

  return {
    post: newPost,
  };
};
```
Lastly, you could use a `form` to call the server action, like this:
```tsx
"use client";

export async function PostCreationForm() {
  const [state, formAction] = useFormState(createPostAction, {
    errors: {},
  });

  useEffect(() => {
    if (state.post) {
      redirect("/posts");
    }
  }, [state]);

  return (
    <form
      action={formAction}
    >
      {/* ... */}
    </form>
  )
}
```
p.s. If you don't want to use a `form`, you could use `useTransition` hook to call the server action:
```tsx
const [isPending, startTransition] = useTransition();
// ...
<Button
  variant="destructive"
  onClick={() => {
    startTransition(async () => {
      await deletePostAction(null, post.id);
    });
  }}
  disabled={isPending}
>
  Delete
</Button>
```

**Comment by samcx** [comment]
@jack-szeto Thanks for sharing!

In the meantime, I'll be closing this issue. If we issue arises again, feel free to create a new [bug report](https://github.com/vercel/next.js/issues/new?assignees=&labels=template:+bug&projects=&template=1.bug_report.yml) so we can take a closer look!

**Comment by 6bangs** [comment]
This is definitely not fixed; I've reproduced it with a minimal case here: https://github.com/6bangs/nextjs-bug

This is just `npx create-next-app@latest` with the following changes:

`Link` tag on the main page:
```html
          Link to bug:&nbsp;
          <Link href="/bug/3" prefetch={false}>click here</Link>
```

Dynamic route with optional catch-all segment at `app/bug/[[...id]]/page.tsx`:
```typescript
export const dynamic = "force-dynamic";

type Props = {
  params: {
      id: string;
  }
}
const BugPage = async ({params: {id}}: Props) => {
  console.log(`Bug page executing for ID ${id}`);
  return (
    <p>Testing.</p>
  )
}

export default BugPage;
```

Repro steps:

1. Clone the repo and run it
2. Click the link to the bug in the upper-left
3. In the console, you'll see `Bug page executing for ID 3`
4. Click the browser back button
5. Click the link again
6. The page is **not executed again** so no message is logged to the console.

No combination of `dynamic` or `revalidate` resolves the issue. I've also tried using `revalidatePath` without success.

IMO this ticket should be reopened rather than creating a new one; there are ~30 others who have run into the issue and a lot of useful comments.

**Comment by samcx** [comment]
@6bangs Thanks for sharing <picture data-single-emoji=":repro:" title=":repro:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/repro/5cea53b5532825b4.png" alt=":repro:" align="absmiddle"></picture>! I've reopened the issue.

Will be taking a look soon at the <picture data-single-emoji=":repro:" title=":repro:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/repro/5cea53b5532825b4.png" alt=":repro:" align="absmiddle"></picture>!

**Comment by jack-szeto** [comment]
@6bangs 
I've been looking at your example and I have a few questions. I understand how it might be confusing when a page doesn't rerender despite having `dynamic` or `revalidate` set.

However, in your example repo, I noticed that nothing changes when you revisit the page. Shouldn't it be cached in this case?

Also, I'm wondering if there are any specific cases where you'd want a dynamic page that doesn't make use of the `unstable_cache` and `tag` functions? I'm just trying to see if there are any scenarios that I might not have considered.

**Comment by 6bangs** [comment]
For my AI use case:

- The first time a user visits the page for a given ID, messages will not exist in the conversation. The visit triggers that asynchronous process.
- The second time the user visits, the page should be executed again and the user should see the messages from the last visit. Since that doesn't happen, the process starts over. The conversation repeats as many times as the user visits the page before the cache is invalidated.

This is my situation, but the bug would be noticed for any asynchronous behavior that modifies the state associated with the segment.

**Comment by samcx** [comment]
@6bangs This is actually separate from the issue that was originally reported since there is no `fetch()` in your example—the `export const dynamic = "force-dynamic";` will not affect the [Router Cache](https://nextjs.org/docs/app/building-your-application/caching#router-cache). You'll see it log again after [30s](https://nextjs.org/docs/app/building-your-application/caching#duration-3). So your example is working as expected.

There will be an upcoming API (`staletime`) that will let you customize this, but we are still listening to feedback on this! We also need to improve the observability here, as it is not immediately obvious when the Router Cache was used.

- https://github.com/vercel/next.js/discussions/54075
- https://github.com/vercel/next.js/discussions/54075#discussioncomment-6754339

**Comment by 6bangs** [comment]
Yup, `staletime` would work for my use case. Any ETA on that? For now we're stuck with a hacky workaround (appending a cache-busting parameter to the URI).

**Comment by run4w4y** [comment]
Hello! I'm having a similar presumably server-side caching issue (I'm assuming so since the issue persists across different browsers and devices). I have a page for which I get the contents with fetch in a server component, Next version I'm using is `14.1.0`, the route segment config for the whole layout is as follows: 
```javascript
export const dynamic = 'force-dynamic'
export const fetchCach = 'force-no-store'
export const revalidate = false
```
The issue I'm having is: when the content is updated on the backend (the one I'm fetching from in the RSC) the page does show the new content on hard navigation, but when the same page is opened using client-side navigation all I'm seeing is stale content. 
For now I just stick to using hard navigation for all of the links, but hopefully I could get some help with the issue

**Comment by samcx** [comment]
@6bangs No timetable for this yet but hopefully soon!

@run4w4y Can you file a new [bug report](https://github.com/vercel/next.js/issues/new?assignees=&labels=template:+bug&projects=&template=1.bug_report.yml) so we can take a closer look?

**Comment by run4w4y** [comment]
@samcx Thank you for your response! I was thinking this might be related to the issue discussed above, I'm sorry if it is not. In that case I'll try to come up with a smaller repro tomorrow and open up another bug report

**Comment by wieruszewski** [comment]
I am having the same issue with next.js v  `14.0.3`.  I am using a fetch GET request and the response always has `X-Nextjs-Cache: HIT` even with hard refresh. It works fine when I use a dev mode and this issue occurs in a production build.
The only thing which worked was `router.refresh()`. It seems that it is a router cache, but I thought that this cache will be used when using e.g. `Link` components, but not with hard refresh of a page in browser. Is that correct behavior?

**Comment by mirano-galijasevic** [comment]
> @6bangs No timetable for this yet but hopefully soon!
> 
> @run4w4y Can you file a new [bug report](https://github.com/vercel/next.js/issues/new?assignees=&labels=template:+bug&projects=&template=1.bug_report.yml) so we can take a closer look?

More than a month later, still no timetable for fixing this? This had been reported 8 month ago.
Please let me know if you are going to fix this or not. I do not want to use 10 hacks to avoid this in different situations, just because somebody is dead set on doing things the wrong way. Just tell me if you are not going to fix this, so that I can move to some other framework, since you seem to be on the path to ruin this. Thank you.

**Comment by samcx** [comment]
@mirano-galijasevic As far as I'm aware, there were different issues being reported in this thread, and there was no <picture data-single-emoji=":repro:" title=":repro:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/repro/5cea53b5532825b4.png" alt=":repro:" align="absmiddle"></picture> given for the original issue post.

We can't reliably debug and fix and issue without a reproduction. Unfortunately, that's just the case here (according to the original post). There are other bugs and edge cases we're fixing constantly (ones that are older than 8 months or younger), so not saying we're not fixing this, and I'm fully present of new notifications coming into this feed.

**Comment by kevincobain2000** [comment]
Thanks for watching this issue @samcx 
yes there are different issues reported here while all of those can be fixed by fixing the revalidate to be transparent and work well.

Which is

`revalidate:0` 0 means 0 - no cache.

And in support of @mirano-galijasevic - I feel your pain and just due to this issue, many times we had to choose a framework other than Next14.

**Comment by wieruszewski** [comment]
In my case the problem was that the generated route was static content:
![image](https://github.com/vercel/next.js/assets/6325279/d1629a45-693d-4682-8832-253a971dac54)
In that route I only had a `GET` method defined and in this instance every `fetch` request to this route was cached regardless of my config (even with a cache boosting parameter). If I add there any dummy HTTP methods there like `POST` or `PUT` the build will generate this endpoint as dynamic and it works as expected. Or, I could use an `edge` runtime too.

**Comment by samcx** [comment]
@kevincobain2000 Not able to reproduce myself.

![CleanShot 2024-02-29 at 13 23 55@2x](https://github.com/vercel/next.js/assets/28912696/a46c3a13-a0ec-44e0-b8f4-3b7e20fd3eb7)

(`/app/page.tsx`)

```
export default async function Page() {
  const data = await getData();
  return (
    <div>
      <div>{JSON.stringify(data)}</div>
    </div>
  );
}

async function getData() {
  const res = await fetch("https://catfact.ninja/fact", {
    next: { revalidate: 0 },
  });

  if (!res.ok) {
    // This will activate the closest `error.js` Error Boundary
    throw new Error("Failed to fetch data");
  }

  return res.json();
}
```

(`/next.config.mjs`)

```
/**
 * @type {import('next').NextConfig}
 */
const nextConfig = {
  reactStrictMode: true,
  logging: {
    fetches: {
      fullUrl: true,
    },
  },
}

export default nextConfig
```

**Comment by kevincobain2000** [comment]
@samcx 
In your same example above you are refreshing the page and hitting the same url. That is why there is no cache hit.

If you do a soft navigation then it doesn't hit.

```
npx create-next-app -e https://github.com/nextui-org/next-app-template
```

**page.tsx**

```
async function getData() {
    const res = await fetch("https://catfact.ninja/fact", {
      next: { revalidate: 0 },
    });

    if (!res.ok) {
      // This will activate the closest `error.js` Error Boundary
      throw new Error("Failed to fetch data");
    }

    return res.json();
  }

export default async function Home() {
    const data = await getData();
	return (
		<section className="flex flex-col items-center justify-center gap-4 py-8 md:py-10">
            <div>{JSON.stringify(data)}</div>
... 
and so on
```

```
 GET / 200 in 246ms
 │ GET https://catfact.ninja/fact 200 in 211ms (cache: SKIP)
 │  │  Cache missed reason: (revalidate: 0)
 ⚠ Unsupported metadata themeColor is configured in metadata export. Please move it to viewport export instead.
 ⚠ Unsupported metadata themeColor is configured in metadata export. Please move it to viewport export instead.
 GET /robots.txt?1709255277972 404 in 35ms
 GET /about?_rsc=acgkz 200 in 26ms
 ⚠ Unsupported metadata themeColor is configured in metadata export. Please move it to viewport export instead.
 GET /about?_rsc=13o5j 200 in 17ms
 GET /?_rsc=1mruw 200 in 15ms <--- 
```

Reproduce

1. Open localhost:3000
2. View the data
3. Click About
4. Click Home (You see same data and it doesn't make another API call)

**Comment by samcx** [comment]
@kevincobain2000 Just verified what you're seeing.

This is the same thing as mentioned above—the fetch isn't being cached, but there's a separate Router Cache that's caching in-memory. Sounds kind of unintuitive that that we're caching the fetch through the Router Cache that wasn't cached on the server, but that's because one is done on the Client versus the Server.

https://nextjs.org/docs/app/building-your-application/caching#overview

Also to clarify, this isn't something we just created for App Router—there also was a similar Router Cache in Pages Router. You just wouldn't come across the need to bypass/bust that cache (e.g., you on-demand ISR, and want to bust that route's cache too on the client).

Besides the hacky solution above, I think what we need is the Staletime API (modifying the Router Cache timings) which was mentioned above, and a more ergonomic way to possibly opt-out the Router Cache by page. <picture data-single-emoji=":frog-eyes:" title=":frog-eyes:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/frog-eyes/558eb9b0fd306c97.png" alt=":frog-eyes:" align="absmiddle"></picture>

**Comment by kevincobain2000** [comment]
@samcx Thanks for verifying.

Understand what you said, you have put it nicely.

Yes, there needs to be a simple setting that can turn this cache on, off, or for duration.
The reason it has become a big issue, is that most users believe that `revalidate` was for this purpose. And not all users want a router cache, especially when navigation expects data change - trading apps, admin panels, or any other case.

**Comment by buuhvprojects** [comment]
> This is definitely not fixed; I've reproduced it with a minimal case here: https://github.com/6bangs/nextjs-bug
> 
> This is just `npx create-next-app@latest` with the following changes:
> 
> `Link` tag on the main page:
> 
> ```
>           Link to bug:&nbsp;
>           <Link href="/bug/3" prefetch={false}>click here</Link>
> ```
> 
> Dynamic route with optional catch-all segment at `app/bug/[[...id]]/page.tsx`:
> 
> ```ts
> export const dynamic = "force-dynamic";
> 
> type Props = {
>   params: {
>       id: string;
>   }
> }
> const BugPage = async ({params: {id}}: Props) => {
>   console.log(`Bug page executing for ID ${id}`);
>   return (
>     <p>Testing.</p>
>   )
> }
> 
> export default BugPage;
> ```
> 
> Repro steps:
> 
> 1. Clone the repo and run it
> 2. Click the link to the bug in the upper-left
> 3. In the console, you'll see `Bug page executing for ID 3`
> 4. Click the browser back button
> 5. Click the link again
> 6. The page is **not executed again** so no message is logged to the console.
> 
> No combination of `dynamic` or `revalidate` resolves the issue. I've also tried using `revalidatePath` without success.
> 
> IMO this ticket should be reopened rather than creating a new one; there are ~30 others who have run into the issue and a lot of useful comments.

Thanks, use  prefetch={false} in <Link> resolve the problem for me.
My problem was that when change de page content using <link>, the browser doesnt change de document.title in next 13, so it was necessary reload de page to see a new title.

**Comment by BrentJohnsonNz** [comment]
I've just chased down an error where SSR data fetching ceased for a specific simple URL.  My error was caused by the router keeping an entry permanently in the router.js "Server Data Cache" (this.sdc/inflightCache).  The error is caused by executing multiple rapid "router.push(" calls to pages with SSR.  If the first router.push data fetch call is cancelled at exactly the right time (just before it's deleted from the inflight cache) the data gets stuck in the inflight request cache and will be served for all subsequent navigations to the SSR page.

The symptom for the error in this scenario will be a client-side console error "Loading initial props cancelled".

The "inflightCache" appears to only be active for production builds.  In my case it only manifested when running end-to-end tests against staging that were navigating very rapidly between pages.

**Comment by samcx** [comment]
@kevincobain2000 Staletime API (experimental) is available now → https://nextjs.org/docs/app/api-reference/next-config-js/staleTimes.

In the meantime, I will be closing this issue for the time being—feel free to create a new [bug report](https://github.com/vercel/next.js/issues/new?assignees=&labels=template:+bug&projects=&template=1.bug_report.yml) if this is an issue un-related to router cache.

**Comment by kevincobain2000** [comment]
Thanks @samcx 
I ll check this out.

**Comment by 6bangs** [comment]
@samcx, I see this:

> The dynamic property is used when the prefetch prop on Link is left unspecified.
> The static property is used when the prefetch prop on Link is set to true, or when calling [router.prefetch](https://nextjs.org/docs/app/building-your-application/caching#routerprefetch).

What about in my example where prefetch is false?

**Comment by samcx** [comment]
@6bangs good callout! I have created a <picture data-single-emoji=":pr:" title=":pr:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/pr/876af7e7d65e1595.png" alt=":pr:" align="absmiddle"></picture> to fix this → https://github.com/vercel/next.js/pull/65252.

**Comment by 6bangs** [comment]
Great, thanks!
